tests/Microsoft.Identity.Test.Unit.net45/CoreTests/OAuth2Tests/ClaimsTest.cs

```diff
@@ -2,19 +2,256 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Identity.Client;
 using Microsoft.Identity.Client.Core;
 using Microsoft.Identity.Client.Internal;
 using Microsoft.Identity.Client.OAuth2;
 using Microsoft.Identity.Client.TelemetryCore;
+using Microsoft.Identity.Client.UI;
+using Microsoft.Identity.Test.Common.Core.Helpers;
 using Microsoft.Identity.Test.Common.Core.Mocks;
+using Microsoft.Identity.Test.Common.Mocks;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Identity.Test.Unit.CoreTests.OAuth2Tests
 {
     [TestClass]
-    public class ClaimsTest
+    public class ClaimsTest : TestBase
     {
+        // https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1639
+        [TestMethod]
+        public async Task AcquireTokenSilent_BypassesCacheForAccessTokens_IfClaimsUsed_Async()
+        {
+            using (var harness = CreateTestHarness())
+            {
+                Trace.WriteLine("1. Setup an app with a token cache with a valid AT");
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                                            .WithHttpManager(harness.HttpManager)
+                                                                            .BuildConcrete();
+
+                var tokenCacheHelper = new TokenCacheHelper();
+                tokenCacheHelper.PopulateCache(app.UserTokenCacheInternal.Accessor, addSecondAt: false);
+
+                Trace.WriteLine("2. Silent Request without claims returns the AT from the cache");
+                var account = (await app.GetAccountsAsync().ConfigureAwait(false)).Single();
+                var result = await app.AcquireTokenSilent(TestConstants.s_scope, account)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Trace.WriteLine("3. Silent Request + WithClaims does not return an AT from the cache");
+                harness.HttpManager.AddInstanceDiscoveryMockHandler(TestConstants.AuthorityUtidTenant);
+                harness.HttpManager.AddMockHandlerForTenantEndpointDiscovery(TestConstants.AuthorityUtidTenant);
+                harness.HttpManager.AddSuccessTokenResponseMockHandlerForPost(TestConstants.AuthorityUtidTenant);
+
+                result = await app.AcquireTokenSilent(TestConstants.s_scope, account)
```

Comment:
For this test method and following one, how do we assert that step 2 and 3 have hit or bypassed the cache respectively?

Reply:
Hmm, this is a very good question Peter. The way it works is ... if an access token is found in the cache, it is returned imediately. Otherwise, the cache logic is to find a refresh token and make an Http request to AAD called the "refersh_token" grant. 

So if the cache is hit -> no extra network calls occur
If the cache is not hit -> 3 network calls occur (2 for discovery, 1 for /token)

That `using` at line 31 ensures that all mock handlers added (lines 48-50) are actually consumed.

Reply:
But I agree that this is too subtle, let me try to find a way to make these assertions in a more obvious way.

Reply:
Ideally we'd want to mock ICacheSessionManager, but exposing that is pretty tricky. I think I will let these tests be as they are for now and add more comments and assertions to explain what they do.

Reply:
Ok, I understand. I'll take a look at that harness class. Yea, it's not super straightforward. Agree that for now, it's fine, since it's consistent. For the future, we should maybe think about improving this flow.